### PR TITLE
DAOS-2308: Convert btree test to CMOCKA model to produce XML output.

### PR DESCRIPTION
**Pull request by rpadma2**
Signed-off-by: Padmanabhan <ravindran.padmanabhan@intel.com>

### CODE REVIEW

**Comment by jolivier23** [comment]
Looks like there are conflicts with master to resolve

File: src/common/tests/btree.c [hunk 1]
```diff
@@ -5,7 +5,7 @@
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- *    http://www.apache.org/licenses/LICENSE-2.0
+ *	  http://www.apache.org/licenses/LICENSE-2.0
```
Comment (jolivier23) [inline src/common/tests/btree.c]: is this change intended?

File: src/common/tests/btree.c [hunk 2]
```diff
@@ -66,7 +83,7 @@ struct ik_rec {
 
 #define IK_TREE_CLASS	100
 #define POOL_NAME "/mnt/daos/btree-test"
-#define POOL_SIZE ((1024 * 1024  * 1024ULL))
+#define POOL_SIZE ((1024 * 1024	 * 1024ULL))
```
Comment (jolivier23) [inline src/common/tests/btree.c]: is this a tab?  spaces here are fine

Comment (rpadma2) [reply src/common/tests/btree.c]: Removed the tabs.

File: src/common/tests/btree.c (hunk 1 above)
Comment (rpadma2) [reply src/common/tests/btree.c]: Removed the tabs.

File: src/common/tests/btree_direct.c [hunk 3]
```diff
@@ -496,15 +523,14 @@ sk_btr_kv_operate(enum sk_btr_opc opc, char *str, bool verbose)
 		case BTR_OPC_DELETE_RETAIN:
 			rc = dbtree_delete(sk_toh, &key_iov, &rec_mmid);
 			if (rc != 0) {
-				D_ERROR("Failed to delete %s\n", key);
-				return -1;
+				sprintf(outbuf, "Failed to delete %s\n", key);
```
Comment (ravalsam) [inline src/common/tests/btree_direct.c]: sprintf can be changed to asprintf for printing any data length.

File: src/common/tests/btree_direct.c [hunk 4]
```diff
@@ -519,8 +545,8 @@ sk_btr_kv_operate(enum sk_btr_opc opc, char *str, bool verbose)
 			daos_iov_set(&val_iov, NULL, 0); /* get address */
 			rc = dbtree_lookup(sk_toh, &key_iov, &val_iov);
 			if (rc != 0) {
-				D_ERROR("Failed to lookup %s\n", key);
-				return -1;
+				sprintf(outbuf, "Failed to lookup %s\n", key);
```
Comment (ravalsam) [inline src/common/tests/btree_direct.c]: Move this to standard utils so can be called with single line in current file and can be used in other unit test files.

File: src/common/tests/btree_direct.c (hunk 3 above)
Comment (rpadma2) [reply src/common/tests/btree_direct.c]: I believe asprintf is not suitable for this place.
Reason: we will start having memory leaks since the allocated buffer will not be released due to fail_msg terminating the program.
eg: char *x;
      asprintf(&x, "Test failed\n");
      fail_msg("%s" , &x);  ---> we cannot free x before the fail_msg.

File: src/common/tests/btree_direct.c (hunk 4 above)
Comment (rpadma2) [reply src/common/tests/btree_direct.c]: As discussed this may have to rethink this method. The input for the function should handle %s, %s, %u or DF_64U types. We will not implement this change for now.

File: src/common/tests/btree_direct.c (hunk 3 above)
Comment (ravalsam) [reply src/common/tests/btree_direct.c]: That is true. I think that can be the issue with sprintf too, right ?

Comment (rpadma2) [reply src/common/tests/btree_direct.c]: In the present implementation, we are allocating fixed array outbuf[64] which is in stack. sprintf is using this and when program terminates all stack data is given back. asprintf is creating data in heap which needs to be freed before program terminates.

Comment (ravalsam) [reply src/common/tests/btree_direct.c]: ok make sense
